src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs

```diff
@@ -0,0 +1,212 @@
+    }
+
+    protected override IEnumerator<QueueItem> ConvertCountersEnumerator(DocumentsOperationContext context,
+        IEnumerator<CounterGroupDetail> counters, string collection)
+    {
+        throw new NotSupportedException("Counters aren't supported by Queue ETL");
+    }
+
+    protected override IEnumerator<QueueItem> ConvertTimeSeriesEnumerator(DocumentsOperationContext context,
+        IEnumerator<TimeSeriesSegmentEntry> timeSeries, string collection)
+    {
+        throw new NotSupportedException("Time series aren't supported by Queue ETL");
+    }
+
+    protected override IEnumerator<QueueItem> ConvertTimeSeriesDeletedRangeEnumerator(DocumentsOperationContext context,
+        IEnumerator<TimeSeriesDeletedRangeItem> timeSeries, string collection)
+    {
+        throw new NotSupportedException("Time series aren't supported by Queue ETL");
+    }
+
+    protected override EtlTransformer<QueueItem, QueueWithEvents, EtlStatsScope, EtlPerformanceOperation> GetTransformer(
+        DocumentsOperationContext context)
+    {
+        return new QueueDocumentTransformer(Transformation, Database, context, Configuration);
+    }
+
+    protected override int LoadInternal(IEnumerable<QueueWithEvents> items, DocumentsOperationContext context, EtlStatsScope scope)
+    {
+        int count = 0;
+
+        switch (Configuration.Connection.Provider)
+        {
+            case QueueProvider.Kafka:
+                count = ProcessKafka(items, Configuration, context);
+                break;
+            case QueueProvider.RabbitMq:
+                ProcessRabbitMq();
+                break;
+        }
+
+        return count;
+    }
+
+    protected override EtlStatsScope CreateScope(EtlRunStats stats)
+    {
+        return new EtlStatsScope(stats);
+    }
+
+    private int ProcessKafka(IEnumerable<QueueWithEvents> items, QueueEtlConfiguration queueEtlConfiguration,
+        DocumentsOperationContext context)
+    {
+        int count = 0;
+
+        using (var jsonSerializer = new BlittableEventBinaryFormatter(context))
+        {
+            var events = new List<KafkaMessageEvent>();
+
+            foreach (QueueWithEvents queueItem in items)
+            {
+                var topicName = queueItem.Name;
+
+                foreach (var insertItem in queueItem.Inserts)
+                {
+                    var eventMessage = new CloudEvent
+                    {
+                        Id = DefaultCloudEventId,
+                        DataContentType = "application/json",
+                        Type = string.IsNullOrWhiteSpace(queueItem.Type) ? DefaultType : queueItem.Type,
+                        Source = new Uri(string.IsNullOrWhiteSpace(queueItem.Source) ? DefaultSource : queueItem.Source),
+                        Data = insertItem.TransformationResult
+                    };
+
+                    eventMessage.SetPartitionKey(insertItem.ChangeVector);
```

Comment:
We already confirmed that it's not related to idempotent kafka producer. So a better choice would be to use source document id - this way for example order and it's orders if put into separate topics would be placed in the same partition what I believe is the demanded behavior.

Also we have to make it configurable to the user because it's very important. See my above comment - we could set an optional parameter of `loadTo()` method where they could set the options to an event - `loadToUsers({ ... json message ... }. { partitionKey: '....' })`

Reply:
Discussed directly